File: backend/internal/dto/pesagem.go
package dto

import (
	"time"

	"github.com/google/uuid"
)

// PesagemItemRequest representa um item (amostra) dentro de uma pesagem.
type PesagemItemRequest struct {
	Quantidade int     `json:"quantidade" validate:"required,gt=0"`
	Peso       float64 `json:"peso" validate:"required,gt=0"`
}

// CreatePesagemRequest representa o payload de criacao de uma pesagem.
type CreatePesagemRequest struct {
	Data  string               `json:"data" validate:"required"` // formato: YYYY-MM-DD
	Items []PesagemItemRequest `json:"items" validate:"required,min=1,dive"`
}

// PesagemItemResponse representa um item na resposta de pesagem.
type PesagemItemResponse struct {
	ID         uuid.UUID `json:"id"`
	Quantidade int       `json:"quantidade"`
	Peso       float64   `json:"peso"`
	PesoMedio  float64   `json:"peso_medio"`
}

// PesagemResponse representa a resposta de uma pesagem.
type PesagemResponse struct {
	ID              uuid.UUID             `json:"id"`
	LoteID          uuid.UUID             `json:"lote_id"`
	Data            time.Time             `json:"data"`
	QuantidadeTotal int                   `json:"quantidade_total"`
	PesoTotal       float64               `json:"peso_total"`
	PesoMedio       float64               `json:"peso_medio"`
	Items           []PesagemItemResponse `json:"items,omitempty"`
	CreatedAt       time.Time             `json:"created_at"`
}

File: flutter_app/lib/features/lote_detail/domain/entities/pesagem.dart
/// Registro de pesagem de um lote.
///
/// Cada pesagem contem um ou mais [PesagemItem], representando
/// amostras individuais de peso coletadas no campo.
///
/// Campos mapeados a partir do endpoint GET /lotes/{id}/pesagens
/// e do DTO PesagemResponse do backend Go.
class Pesagem {
  final String id;
  final String loteId;
  final String data;
  final int quantidadeTotal;

  /// Peso total em gramas.
  final double pesoTotal;

  /// Peso medio em gramas.
  final double pesoMedio;

  /// Itens individuais da pesagem (campo `items` do backend).
  final List<PesagemItem> items;

  final String createdAt;

  const Pesagem({
    required this.id,
    required this.loteId,
    required this.data,
    required this.quantidadeTotal,
    required this.pesoTotal,
    required this.pesoMedio,
    required this.items,
    required this.createdAt,
  });

  factory Pesagem.fromJson(Map<String, dynamic> json) {
    return Pesagem(
      id: json['id'] as String,
      loteId: json['lote_id'] as String,
      data: json['data'] as String,
      quantidadeTotal: (json['quantidade_total'] as num).toInt(),
      pesoTotal: (json['peso_total'] as num).toDouble(),
      pesoMedio: (json['peso_medio'] as num).toDouble(),
      items: (json['items'] as List<dynamic>?)
              ?.map(
                (e) => PesagemItem.fromJson(e as Map<String, dynamic>),
              )
              .toList() ??
          const [],
      createdAt: json['created_at'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'lote_id': loteId,
      'data': data,
      'quantidade_total': quantidadeTotal,
      'peso_total': pesoTotal,
      'peso_medio': pesoMedio,
      'items': items.map((e) => e.toJson()).toList(),
      'created_at': createdAt,
    };
  }
}

/// Item individual de uma pesagem.
///
/// Representa uma amostra: quantidade de aves pesadas e o peso
/// registrado em gramas.
class PesagemItem {
  final String id;

  /// Quantidade de aves nesta amostra.
  final int quantidade;

  /// Peso total desta amostra em gramas.
  final double peso;

  /// Peso medio desta amostra em gramas.
  final double pesoMedio;

  const PesagemItem({
    required this.id,
    required this.quantidade,
    required this.peso,
    required this.pesoMedio,
  });

  factory PesagemItem.fromJson(Map<String, dynamic> json) {
    return PesagemItem(
      id: json['id'] as String,
      quantidade: (json['quantidade'] as num).toInt(),
      peso: (json['peso'] as num).toDouble(),
      pesoMedio: (json['peso_medio'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'quantidade': quantidade,
      'peso': peso,
      'peso_medio': pesoMedio,
    };
  }
}
